hls.js/index.d.ts

```diff
@@ -0,0 +1,523 @@
+        enableWorker?: boolean;
+        enableSoftwareAES?: boolean;
+        manifestLoadingTimeOut?: number;
+        manifestLoadingMaxRetry?: number;
+        manifestLoadingRetryDelay?: number;
+        manifestLoadingMaxRetryTimeout ?: number;
+        levelLoadingTimeOut?: number;
+        levelLoadingMaxRetry?: number;
+        levelLoadingRetryDelay?: number;
+        levelLoadingMaxRetryTimeout?: number;
+        fragLoadingTimeOut?: number;
+        fragLoadingMaxRetry?: number;
+        fragLoadingRetryDelay?: number;
+        fragLoadingMaxRetryTimeout?: number;
+        fragLoadingLoopThreshold?: number;
+        startFragPrefech?: boolean;
+        appendErrorMaxRetry?: number;
+        loader?: Loader;
+        fLoader?: Loader;
+        pLoader?: Loader;
+        xhrSetup?: (xhr?: XMLHttpRequest, url?: string) => void;
+        fetchSetup?: (context?: LoaderContext, params?: any) => void;
+        abrController?: any;
+        timelineController?: any;
+        enableCEA708Captions?: boolean;
+        stretchShortVideoTrack?: boolean;
+        forceKeyFrameOnDiscontinuity?: boolean;
+        abrEwmaFastLive?: number;
+        abrEwmaSlowLive?: number;
+        abrEwmaFastVoD?: number;
+        abrEwmaSlowVoD?: number;
+        abrEwmaDefaultEstimate?: number;
+        abrBandWidthFactor?: number;
+        abrBandWidthUpFactor?: number;
+    }
+
+    export interface ErrorEvent {
+        type: ErrorType;
+        details: ErrorDetails;
+        fatal: boolean;
+    }
+
+    export interface QualityLevel {
+        attrs: any; // todo
+        details?: any; // todo
+
+        audioCodec: string;
+        videoCodec: string;
+        bitrate: number;
+        url: string[];
+        urlId: number;
+        name: string;
+        width: number;
+        height: number;
+    }
+
+    export interface Instance {
+        /**
+         * Calling this method will:
+         *  - bind videoElement and hls instance;
+         *  - create MediaSource and set it as video source
+         *  - once MediaSource object is successfully created;
+         *    MEDIA_ATTACHED event will be fired.
+         */
+        attachMedia(element: HTMLVideoElement): void;
+
+        /**
+         * Calling this method will:
+         *  - unbind VideoElement from hls instance;
+         *  - signal the end of the stream on MediaSource
+         *  - reset video source (video.src = '')
+         */
+        detachMedia(): void;
+
+        /**
+         * Should be called to free used resources and destroy hls context.
+         */
+        destroy(): void;
+
+        /**
+         * Starts loading video data from the manifest URL.
+         */
+        loadSource(url: string): void;
+
+        /**
+         * on attaches a listener an en event.
+         */
+        on(event: 'hlsError', handler: (data: ErrorEvent) => void): Instance;
+        on(event: EventName, handler: (...data: any[]) => void): Instance;
```

Comment:
> on(event: EventName, handler: (...data: any[]) => void): Instance;

It seem like that this `handler` argument is following:

```js
hls.on(Hls.Events.LEVEL_LOADED, function(event, data) {
  var level_duration = data.details.totalduration;
});
```

- https://github.com/dailymotion/hls.js/blob/master/doc/API.md#runtime-events
- https://github.com/dailymotion/hls.js/blob/2f91b1e7b6fc38c21e1300be66d3c2864f2bd610/src/event-handler.js#L64-L69